wire-ios-request-strategy/Sources/Synchronization/QuickSyncObserver.swift

```diff
@@ -39,34 +61,59 @@ public final class QuickSyncObserver: QuickSyncObserverInterface {
         self.context = context
         self.applicationStatus = applicationStatus
         self.notificationContext = notificationContext
+
+        setupObservation()
     }
 
-    public func waitForQuickSyncToFinish() async {
-        if await quickSyncHasCompleted() {
+    private func setupObservation() {
+        notificationCenter
+            .publisher(for: .didStartDecryptingEventsNotification)
+            .receive(on: decryptionQueue)
+            .sink { [weak self] _ in
+                self?.decryptionState = .inProgress
+            }
+            .store(in: &cancellables)
+
+        notificationCenter
+            .publisher(for: .didStopDecryptingEventsNotification)
+            .receive(on: decryptionQueue)
+            .sink { [weak self] _ in
+                self?.decryptionState = .done
+            }
+            .store(in: &cancellables)
+    }
+
+    public func waitForDecryptionOfEventsToFinish() async {
+        if await quickSyncHasCompleted() || finishedDecrypting {
```

Comment:
I changed again the implementation to early return if all quick sync has been finished